src/calling/haplotypes.rs

```diff
@@ -177,49 +184,93 @@ impl KallistoEstimates {
 pub(crate) struct Variant(#[deref] String);
 
 #[derive(Derefable, Debug, Clone, PartialEq, Eq)]
-pub(crate) struct HaplotypeVariants(#[deref] HashMap<Variant, Vec<Haplotype>>);
+pub(crate) struct HaplotypeVariants(#[deref] HashMap<Variant, BitVec>);
 
 impl HaplotypeVariants {
-    pub(crate) fn new(haplotype_variants: &mut bcf::Reader) -> Result<Vec<HaplotypeVariants>> {
-        let mut vec_haplotype_variants = Vec::new();
+    pub(crate) fn new(haplotype_variants: &mut bcf::Reader) -> Result<Self> {
+        let mut candidate_variants = HashMap::new();
         for record_result in haplotype_variants.records() {
-            let record = record_result.expect("Failed to read record");
+            let record = record_result?;
 
             //store the variant
             let mut variant = String::new();
             variant.push_str(&format!("{}{}", record.contig(), "_"));
             let pos = record.pos().to_string();
             variant.push_str(&format!("{}{}", pos, "_"));
             for allele in record.alleles() {
-                for c in allele {
-                    let base = char::from(*c).to_string();
-                    variant.push_str(&format!("{}{}", base, "_"));
+                if allele.len() == 1 {
+                    for c in allele {
+                        let base = char::from(*c).to_string();
+                        variant.push_str(&format!("{}{}", base, "_"));
+                    }
+                } else {
+                    //the case for structural variants
+                    for c in allele {
+                        let base = char::from(*c).to_string();
+                        variant.push_str(&base);
+                    }
+                    variant.push_str("_")
                 }
             }
 
             //store the haplotypes that carry the variant
             let header = record.header();
-            let gts = record.genotypes().expect("Error reading genotypes"); //genotypes of all samples
-            let sample_count = usize::try_from(record.sample_count()).unwrap();
-            let sample = (0..sample_count).collect::<Vec<_>>();
-            let mut haplotypes: Vec<Haplotype> = Vec::new();
+            let gts = record.genotypes()?; //genotypes of all samples
 
-            for (sample_index, mut x) in sample.iter().zip(header.samples().into_iter()) {
-                for gta in gts.get(*sample_index).iter() {
+            let mut bv: BitVec<usize> = BitVec::new();
+            for (sample_index, x) in header.samples().into_iter().enumerate() {
+                for gta in gts.get(sample_index).iter() {
```

Comment:
I am unsure whether this will give you the correct order in the resulting bitvector. If I get your decision above right, you chose to take the kallisto file as the reference order for the haplotypes. Then, I think here you will have to ensure that the bitvector is filled in the same order. Otherwise you won't be able to use the same index later on to access both data.